### DISCUSSION TITLE
IndexedBitmap?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16264] [Q&A] [answered]
Is it possible to create a fast, low memory footprint, IndexedBitmap wrapper around Bitmap?
Maybe Writeable, Cropped, and Renderable versions as well?

**Comment by maxkatz6** [comment]
What do you mean by "indexed" bitmap?

In general, any "writeable" bitmaps won't be as fast, since they have to stay in CPU RAM, and can't be kept in GPU memory. Or, otherwise, you would need to do expensive bitmap transfers between memories on each edit.

**Reply by Shadowblitz16** [reply]
I mean like mapping a pixel's value to a color on a 2d array of colors.

**Reply by maxkatz6** [reply]
You can build it on top of WriteableBitmap. Locking bitmap on each pixel write.
It's inefficient but will work. Depending on what's your task is.

You should always combine writes to the bitmap in a single `Lock`, avoiding situations like flawed System.Drawing.Bitmap.SetPixel design. But if it's rare to write into your bitmap, it shouldn't be a problem.

**Reply by Shadowblitz16** [reply]
> You can build it on top of WriteableBitmap. Locking bitmap on each pixel write. It's inefficient but will work. Depending on what's your task is.
> 
> You should always combine writes to the bitmap in a single `Lock`, avoiding situations like flawed System.Drawing.Bitmap.SetPixel design. But if it's rare to write into your bitmap, it shouldn't be a problem.

How would I actually do that though? 
I am currently trying to figure out how to calculate RowBytes for a LockedIndexFramebuffer.

**Comment by stevemonaco** [comment]
These problems always have tradeoffs, so you should describe the problem you're actually trying to solve. I have some relevant source, but they don't try to solve a compact storage problem as: 1. GPUs don't work with indexed graphics, AFAIK. 2. my images can fit into memory. Indexed color support (palettes) seems to have been dropped from SkiaSharp quite awhile back, so you can't rely on that.

https://github.com/stevemonaco/AvaloniaDemos contains `RealTimeBitmapAdapter` and `SkiaBitmapAdapter` which are suitable for realtime display, but I wouldn't call them "fast".

https://github.com/stevemonaco/ImageMagitek/tree/main/TileShop.UI/ViewExtenders/Imaging has some adapters for an app that doesn't have realtime display. However, the indexed adapter is highly specific to this app's domain problem because multiple palettes can be present in a single image. I don't recommend supporting that.

**Reply by Shadowblitz16** [reply]
@stevemonaco I need indexed bitmaps because I am making a tile based game in which there will be lots of different recolors for tiles. 
There is also going to be palette lerping between screen transitions.
Having 1000's of different recolors for each tile is just not possible

Normally this would be handled by having a remap shader and a palette swap shader but avalonia doesn't support pixel shaders as far as I know

**Reply by maxkatz6** [reply, marked as answer]
You can use SKBitmap and skia shaders, if you want to go a lower-level.

**Reply by Shadowblitz16** [reply]
k thankyou

**Reply by Shadowblitz16** [reply]
> You can use SKBitmap and skia shaders, if you want to go a lower-level.

I looked up SKBitmap and it looks complicated.
I might go with WriteableBitmap

**Reply by Shadowblitz16** [reply]
So I changed my mind I am going to go with a shader however...

I don't know how to even get a SKBitmap from a RenderTargetBitmap
```cs
public class Surface : RenderTargetBitmap
{
    public Surface(uint width, uint height) : base(new((int)width, (int)height))
    {
        var target = this.CreateDrawingContext();
        var canvas = target.Ge
    }
}
```
